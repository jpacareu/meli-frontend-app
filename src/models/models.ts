import { History } from 'history';

export interface IState {
	items: IItem[];
	searchBox: string;
	categories: string[];
	loadingSearch: boolean;
	detailProduct: {
		item: IDetailItem;
	} | null;
}

export interface IActions {
	search: (...rest: any) => {};
	getProductDetail: (...rest: any) => {};
	loading: (...rest: any) => {};
	received: (...rest: any) => {};
	setState: (...rest: any) => {};
	handleError: (...rest: any) => {};
}

export interface IStateHistory extends IState {
	history: History;
}

export interface IDetailItem {
	title: string;
	price: IPrice;
	picture: string;
	condition: ICondition;
	free_shipping: boolean;
	sold_quantity: number;
	description: string;
}

export interface IItem {
	id: string;
	title: string;
	price: IPrice;
	picture: string;
	condition: ICondition;
	free_shipping: boolean;
}

export type ICondition = 'new' | 'used';

export interface IPrice {
	currency: string;
	amount: number;
	decimals: number;
}

export interface IStore {
	Provider: any;
	connect: () => any;
	actions: IActions;
}
