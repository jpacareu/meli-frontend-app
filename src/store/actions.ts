import axios from 'axios';
import { IState, IActions } from '../models/models';

const baseURL =
	process.env.NODE_ENV === 'production'
		? 'https://meli-backend-app.herokuapp.com'
		: 'http://localhost:4000';

const axiosInstance = axios.create({
	baseURL: baseURL + '/api/'
});

export const actionsCreators: IActions = {
	search: async (state: IState, actions) => {
		let resultState: any = {};

		if (state.searchBox === '') return resultState;

		await actions.loading();

		try {
			const result = await axiosInstance.get('/items', {
				params: {
					q: state.searchBox
				}
			});

			const { items, categories } = result.data;
			resultState = { items: items || [], categories: categories || [] };
		} catch (error) {
			await actions.handleError(error);
		} finally {
			await actions.received();
		}

		return resultState;
	},
	getProductDetail: async (state: IState, actions, id: string = '') => {
		let resultState: any = {};
		await actions.loading();
		try {
			const result = await axiosInstance.get(`/items/${id}`);
			const { categories, ...rest } = result.data;
			resultState = {
				detailProduct: rest,
				categories
			};
		} catch (error) {
			await actions.handleError(error);
		} finally {
			await actions.received();
			return resultState;
		}
	},
	loading: () => {
		return { loadingSearch: true };
	},
	received: () => {
		return { loadingSearch: false };
	},
	handleError: (state, actions, error) => {
		// Handle error mock
		return {};
	},
	setState: (state, actions, value = {}) => {
		return { ...value };
	}
};
