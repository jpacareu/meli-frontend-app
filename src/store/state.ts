import { IState } from '../models/models';

export const initialState: IState = {
	items: [],
	searchBox: '',
	categories: [],
	loadingSearch: false,
	detailProduct: null
};
