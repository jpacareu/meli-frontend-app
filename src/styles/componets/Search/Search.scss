
header.header {
    background-color: $main;
}

.search-bar {
    &__container {
        height: 60px;
        display: flex;
        padding: $padding-xs; 
        align-content: center;
    }
    &__input {
        flex-basis: 100%;
        padding: $padding-xs; 
        border-radius: 5px 0 0 5px;
        outline: none;
        font-size: $font-size-18;
    }
    &__button {
        border-radius: 0 5px 5px 0px;
        text-align: center;
    }
    &__image {
        position: absolute;
        top: 12px;
        margin-left: 5px;
    }
}