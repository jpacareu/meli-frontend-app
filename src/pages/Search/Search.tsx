import React, { Fragment } from 'react';
import mercadoLibreLogo from '../../assets/Logo_ML.png';
import searchLogo from '../../assets/ic_Search.png';
import { actions, connect } from '../../store';
import { withRouter, Link } from 'react-router-dom';
import { History } from 'history';

const handleOnSubmit = async (e, search: string, history: History) => {
	e.preventDefault();

	await actions.search();

	history.push({
		pathname: '/items',
		search: `search=${search}`
	});
};

export let Search = props => {
	return (
		<Fragment>
			<div className="offset-md-1 col-md-1 col-2">
				<Link to="/">
					<img
						className="search-bar__image"
						alt="Logo Marcado Libre"
						src={mercadoLibreLogo}
					/>
				</Link>
			</div>
			<div className="col-md-9 col-10">
				<form onSubmit={e => handleOnSubmit(e, props.searchBox, props.history)}>
					<div className="search-bar__container">
						<input
							className="search-bar__input"
							value={props.searchBox}
							onChange={e => actions.setState({ searchBox: e.target.value })}
							placeholder="Nunca dejes de buscar"
						/>
						<button
							className="search-bar__button"
							type="submit"
							onClick={e => handleOnSubmit(e, props.searchBox, props.history)}>
							<img src={searchLogo} alt="Search" />
						</button>
					</div>
				</form>
			</div>
		</Fragment>
	);
};

export default withRouter(connect(({ searchBox }) => ({ searchBox }))(Search));
