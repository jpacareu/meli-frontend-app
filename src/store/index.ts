import createStore from 'react-waterfall';
import { actionsCreators } from './actions';
import { initialState } from './state';

const config = {
	initialState: initialState,
	actionsCreators: actionsCreators
};
const store = createStore(config);

export const { Provider, connect, actions } = store;
