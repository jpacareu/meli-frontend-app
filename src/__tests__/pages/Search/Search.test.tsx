import React from 'react';
import renderer, { ReactTestRenderer } from 'react-test-renderer';
import { IStateHistory } from '../../../models/models';
import { History, createBrowserHistory } from 'history';
import { Search } from '../../../pages/Search/Search';
import { Router, Route } from 'react-router';
import { Provider, actions } from '../../../store';

const props: IStateHistory = {
	loadingSearch: false,
	detailProduct: {
		item: {
			condition: 'new',
			description: 'This is the description',
			free_shipping: true,
			picture: 'http://mywebsite.com/image.jpg',
			sold_quantity: 1000,
			title: 'Title 1',
			price: {
				amount: 1000,
				currency: 'ARS',
				decimals: 0
			}
		}
	},
	categories: ['Category 1', 'Category 2', 'Category 3'],
	items: [],
	searchBox: '',
	history: {} as History
};

let component: JSX.Element | null = null;
let tree: ReactTestRenderer;

describe('<Search />', () => {
	beforeEach(() => {
		const history = createBrowserHistory();
		history.push = () => {};
		component = (
			<Router history={history}>
				<Provider>
					<Route component={Search} exact path="/" />
				</Provider>
			</Router>
		);
		tree = renderer.create(component);
	});

	it('should render without crashing', () => {
		expect(tree.toJSON()).toMatchSnapshot();
	});

	it("should call 'onSubmit'", () => {
		const form = tree.root.findByType('form');
		const event = {
			preventDefault: jest.fn(),
			target: {
				value: 'Value'
			}
		};

		form.props.onSubmit(event);

		expect(event.preventDefault).toHaveBeenCalled();
	});

	it("should call 'onSubmit' when clicking button", () => {
		const form = tree.root.findByType('button');
		const event = {
			preventDefault: jest.fn(),
			target: {
				value: 'Value'
			}
		};
		form.props.onClick(event);

		expect(event.preventDefault).toHaveBeenCalled();
	});

	it("should call 'onChange' when changing text", () => {
		const input = tree.root.findByProps({ className: 'search-bar__input' });
		const event = {
			target: {
				value: 'Value'
			}
		};
		actions.setState = jest.fn();
		input.props.onChange(event);

		expect(actions.setState).toHaveBeenCalledWith({ searchBox: 'Value' });
	});
});
