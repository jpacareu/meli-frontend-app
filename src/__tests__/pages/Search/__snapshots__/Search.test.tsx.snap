// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search /> should render without crashing 1`] = `
Array [
  <div
    className="offset-md-1 col-md-1 col-2"
  >
    <a
      href="/"
      onClick={[Function]}
    >
      <img
        alt="Logo Marcado Libre"
        className="search-bar__image"
        src="Logo_ML.png"
      />
    </a>
  </div>,
  <div
    className="col-md-9 col-10"
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="search-bar__container"
      >
        <input
          className="search-bar__input"
          onChange={[Function]}
          placeholder="Nunca dejes de buscar"
        />
        <button
          className="search-bar__button"
          onClick={[Function]}
          type="submit"
        >
          <img
            alt="Search"
            src="ic_Search.png"
          />
        </button>
      </div>
    </form>
  </div>,
]
`;
